tools/2025-01-gitops-technical-design.md

```diff
@@ -0,0 +1,435 @@
+that are not part of this feature (dependencies). This diagram generally
+treats the components as black boxes. Provide a pointer to a more detailed
+design document, if one exists.
+-->
+
+The Radius Flux Controller reads the Git repository, compiles Bicep files, and creates DeploymentTemplate resources on the cluster. It interacts with the DeploymentTemplate controller to ensure that the desired state defined in the Git repository is accurately reflected in the Kubernetes cluster.
+
+### Architecture Diagram
+<!--
+Provide a diagram of the system architecture, illustrating how different
+components interact with each other in the context of this proposal.
+
+Include separate high level architecture diagram and component specific diagrams, wherever appropriate.
+-->
+
+![Architecture Diagram Part 1](2025-01-gitops-technical-design/rfc-1.png)
+
+![Architecture Diagram Part 2](2025-01-gitops-technical-design/rfc-2.png)
+
+### Detailed Design: Radius Flux Controller
+
+<!--
+This section should be detailed and thorough enough that another developer
+could implement your design and provide enough detail to get a high confidence
+estimate of the cost to implement the feature but isn’t as detailed as the
+code. Be sure to also consider testability in your design.
+
+For each change, give each "change" in the proposal its own section and
+describe it in enough detail that someone else could implement it. Cover
+ALL of the important decisions like names. Your goal is to get an agreement
+to proceed with coding and PRs.
+
+If there are alternatives you are considering please include that in the open
+questions section. If the product has a layered architecture, it's good to
+align these sections with the product's layers. This will help readers use
+their current understanding to understand your ideas.
+
+Discuss the rationale behind architectural choices and alternative options
+considered during the design process.
+-->
+
+The Radius Flux Controller will be implemented as a new controller in the `controller` service. Here is the high-level control loop of the new controller:
+
+1. The controller lists all `GitRepository` resources in the cluster.
+1. For each `GitRepository`, the controller parses the contents of the Git repository.
+1. The controller checks for a `radius-config.yaml` file in the repository. If the file is not present, then the controller skips the repository.
+1. The controller reads the `radius-config.yaml` file and extracts the Bicep compilation instructions, such as the Bicep file path and parameters.
+1. The controller compiles the Bicep file using the Bicep CLI, which is installed on the controller.
+1. The controller creates (or updates) a `DeploymentTemplate` resource on the cluster with the compiled Bicep file as the template.
+1. The controller lists all `DeploymentTemplate` resources in the cluster that have been created from this Git repository and deletes any resources that are not present in the Git repository.
```

Comment:
Why do we need to list all `DeploymentTemplate` resources for the given Git repository? Aren't just the last one and the one before that enough?

Reply:
There may be multiple `DeploymentTemplates` that get created from one git repository (multiple root bicep files). So we need to handle the case where 2 are created on the first commit, then one is removed.